cli script to compute account equity from solana snapshot

ckamm opened the pull request:
don't merge, review only

ckamm committed:
cli script to compute account equity from solana snapshot

ckamm committed:
compute reimbursement amounts

taking into account previous liabs and potentially available token
balances

ckamm committed:
tentative numbers

ckamm committed:
improvements

ckamm committed:
print out mango account address plus owner address

microwavedcola1 commented on cli/src/main.rs:
```diff
@@ -0,0 +1,529 @@
+            "AVAX", "LUNA", "GMT", "USDC",
+        ];
+
+        let available_tokens: [bool; 15] = [
```
why is this an array of size 15 and not 16?

ckamm committed:
take deposits/withdraws after snapshot into account

ckamm committed:
MNGO incentives and MNGO special handling

ckamm commented on cli/src/main.rs:
```diff
@@ -0,0 +1,529 @@
+        //println!("{:?}", ctx.cache.price_cache.iter().map(|c| c.price).collect::<Vec<_>>());
+
+        // TODO: tentative numbers from "Repay bad Debt #2" proposal
+        let available_native_amounts: [u64; 15] = [
```
Index 15 is USDC, where we don't have a specific target number of tokens available. Instead the amount needed for full coverage is an output of the script.

ckamm committed:
fixes after analysis

- disable giving people more tokens, prefer USDC settlement
- settle borrows against non-USDC preferably, to treat delta neutral
  positions more gracefully
- add debug output

ckamm committed:
code cleanup

ckamm committed:
Fix bug and comment

ckamm committed:
Clarify that account skipping is about negative equity

dsemerson commented on cli/src/main.rs:
```diff
@@ -0,0 +1,761 @@
+        // Token prices at time of reimbursement
+        // Note that user equity at snapshot time is computed from the prices from the
+        // mango cache in the snapshot.
+        let reimbursement_prices: [I80F48; 16] = [
```
Are these values going to be hardcoded right before the script gets kicked off? If so I presume multiple people would be verifying the values are correct? Would it be better to reference an oracle at runtime? 

I am a little bit scared that the script is relying on matching up indexes across multiple hardcoded arrays. They do seem to be correctly matched up though, and I suppose the reimbursement prices can have some margin of error.

ckamm replied:
Yes, the prices would be frozen and updated as close to actual reimbursement as we can. I don't think pulling from an oracle is a good idea.

Constants: I agree and had meant to clean that up for a while. Done now.

ckamm committed:
Use one table of token constants instead of multiple

Also, don't merge USDT with USDC early, use the normal equity handling
mechanism for it.

Produces exactly the same output as 721e9b4dc4b2ba2cc9fbd7816338fa3f066f470f

ckamm committed:
Allow for csv and binary output

No change in computed values

ckamm committed:
Update available token balances

Based on what's on 9mM6NfXauEFviFY1S1thbo7HXYNiSWSvwZEhguJw26wY

ckamm committed:
Update reimb prices for more realistic results

Using coingecko as of 20221017 0920 UTC

ckamm committed:
Use snapshot MNGO price for settling MNGO-PERP

Perp positions should be settled at snapshot time. The new MNGO price
should only be used for spot deposits/borrows.

ckamm committed:
Add a distribution-mode=usdc-only flag

In this mode only USDC would be distributed

ckamm committed:
Fixup headers

mschneider committed:
set prices to 10/16/22 close from coingecko

ckamm committed:
Add serum oo token amounts to redistributable tokens

ckamm committed:
Note about reimbursement prices source

ckamm committed:
Make greedy settling and scale down optional

ckamm committed:
Improve token usage debug output

ckamm committed:
Minor debug output improvements

ckamm committed:
Add USDC+MNGO mode

ckamm committed:
Improve settlement matching order

dafyddd committed:
adjust prices for reimbursement to be closing price on 2022-10-17

kronael commented:
Checked our balances for https://docs.google.com/spreadsheets/d/131YQnLUBgatfhqfpf1FQfuRE8SNpdLq1sqraOAyKfLM/edit#gid=1057746867 ... seems legit.  Thanks.

ckamm committed:
Binary output: Only one row per owner, summing accounts
